File: legacy/engine/storage.py
"""Mock S3/GCS object store: 'object storage' as a local directory.

Layout: {root}/{engagement_id}/{hash} — write-once objects, per-engagement
prefix so retention / legal hold / scrubbing operate on one folder.
"""

from __future__ import annotations

import os
import uuid
from pathlib import Path

DEFAULT_ROOT = "mock_s3_gcs"


def payload_ref(engagement_id: int, content_hash: str) -> str:
    return f"{engagement_id}/{content_hash}"


def write_payload(root: str | Path, engagement_id: int, content_hash: str, data: bytes) -> str:
    """Write-once: if the object exists it is never rewritten (content-addressed,
    so identical name means identical bytes)."""
    ref = payload_ref(engagement_id, content_hash)
    path = Path(root) / ref
    if not path.exists():
        path.parent.mkdir(parents=True, exist_ok=True)
        # Unique tmp per writer: concurrent identical-output executions must
        # not share a tmp file (Windows replace-while-open, torn writes). All
        # writers carry identical bytes, so whichever replace lands last wins
        # harmlessly.
        tmp = path.with_name(f"{path.name}.{uuid.uuid4().hex}.tmp")
        try:
            tmp.write_bytes(data)
            os.replace(tmp, path)
        finally:
            if tmp.exists():
                try:
                    tmp.unlink()
                except OSError:
                    pass
    return ref


def read_payload(root: str | Path, ref: str) -> bytes:
    return (Path(root) / ref).read_bytes()

File: backend/src/infrastructure/storage/Storage.ts
import { randomBytes } from 'node:crypto';
import { existsSync, mkdirSync, readFileSync, renameSync, unlinkSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { NotFoundError } from '../../shared/errors/Errors.js';

// Mock S3/GCS payload store. Layout: {root}/{engagement_id}/{hash} — write-once objects with a
// per-engagement prefix so retention / legal hold / scrubbing operate on one folder.

// The string stored in artifacts.payload_ref.
const payloadRef = (engagementId: number, contentHash: string): string => `${engagementId}/${contentHash}`;

// Write-once: if the object exists it is never rewritten (content-addressed, so identical name
// means identical bytes). Atomic rename with a unique tmp file per writer — concurrent
// identical-output executions must not share a tmp file; whichever rename lands last wins harmlessly.
export function writePayload(root: string, engagementId: number, contentHash: string, data: Uint8Array): string {
  const ref = payloadRef(engagementId, contentHash);
  const path = join(root, ref);
  if (!existsSync(path)) {
    mkdirSync(dirname(path), { recursive: true });
    const tmp = `${path}.${randomBytes(16).toString('hex')}.tmp`;
    try {
      writeFileSync(tmp, data);
      renameSync(tmp, path);
    } finally {
      if (existsSync(tmp)) {
        try {
          unlinkSync(tmp);
        } catch {
          // best-effort cleanup
        }
      }
    }
  }
  return ref;
}

export function readPayload(root: string, ref: string): Uint8Array {
  const path = join(root, ref);
  if (!existsSync(path)) {
    throw new NotFoundError(`payload ${ref} not found`, 'payload', ref);
  }
  return readFileSync(path);
}
